Create troubleshoot-high-disk-io.md (#3046)
pingcap/docs: troubleshoot-high-disk-io.md

---
title: Troubleshoot High Disk I/O Usage in TiDB
summary: Learn how to locate and address the issue of high TiDB storage I/O usage.
---

# Troubleshoot High Disk I/O Usage in TiDB

This document introduces how to locate and address the issue of high disk I/O usage in TiDB.

## Check the current I/O metrics

If TiDB's response slows down after you have troubleshot the CPU bottleneck and the bottleneck caused by transaction conflicts, you need to check I/O metrics to help determine the current system bottleneck.

### Locate I/O issues from monitor

> The quickest way to locate I/O issues is to view the overall I/O status from the monitor, such as the Grafana dashboard which is deployed by default by TiDB Ansible and TiUP. The dashboard panels related to I/O include **Overview**, **Node_exporter**, **Disk-Performance**.

Reviewer: ```suggestion
The quickest way to locate I/O issues is to view the overall I/O status from the monitor, such as the Grafana dashboard which is deployed by default by TiDB Ansible and TiUP. The dashboard panels related to I/O include **Overview**, **Node_exporter**, and **Disk-Performance**.
```

#### The first type of monitoring panels

> In `Overview`> `System Info`> `IO Util`, you can see the I/O status of each machine in the cluster. This metric is similar to `util` in the Linux `iostat` monitor. The higher percentage represents higher disk I/O usage:

Reviewer: ```suggestion
In **Overview**> **System Info**> **IO Util**, you can see the I/O status of each machine in the cluster. This metric is similar to `util` in the Linux `iostat` monitor. The higher percentage represents higher disk I/O usage:
```

---
title: Troubleshoot High Disk I/O Usage in TiDB
summary: Learn how to locate and address the issue of high TiDB storage I/O usage.

> category: reference

Reviewer: ```suggestion
category: reference
aliases: ['/docs/dev/troubleshoot-high-disk-io/']
```

> category: reference

Reviewer: ```suggestion
```

- Append log is slow. TiKV Grafana's Raft I/O and append log duration are relatively high. Usually, it is due to slow disk writing. You can check the value of WAL Sync Duration max of RocksDB-raft to confirm, otherwise you may need to report a bug.
- The raftstore thread is busy. TiKV grafana's Raft Propose/propose wait duration is significantly higher than append log duration. Please check the following two points:

    - Is the `store-pool-size` configuration of `[raftstore]` too small (this value is recommended between [1,5], not too large).
    - Is machine's CPU insufficient.

- Apply log is slow. TiKV Grafana's Raft I/O and apply log duration are relatively high, usually accompanied by a relatively high Raft Propose/apply wait duration. The possible situations are as follows:
  
    - The `apply-pool-size` configuration of `[raftstore]` is too small (recommended between [1, 5], not recommended to be too large), Thread CPU/apply cpu is relatively high;
    - The machine's CPU resources are not enough.

>     - Region write hotspot issue, the CPU usage of a single apply thread is relatively high (by modifying the Grafana expression, plus by (instance, name) to see the CPU usage of each thread), temporarily for the hot write of a single Region is no solution, this scene is being optimized recently.

Reviewer: ```suggestion
    - Write hotspot issue of a single Region (Currently, the solution to this issue is still on the way). The CPU usage of a single `apply` thread is high (which can be viewed by modifying the Grafana expression, appended with `by (instance, name)`).
```

>     - It is slow to write RocksDB, and the RocksDB kv/max write duration is relatively high (a single Raft log may contain many kvs. When writing RocksDB, 128 kvs will be written to RocksDB in a batch write, so one apply log may involve multiple RocksDB writes).

Reviewer: ```suggestion
    - Slow write into RocksDB, and `RocksDB kv`/`max write duration` is high. A single Raft log might contain multiple key-value pairs (kv). 128 kvs are written to RocksDB in a batch, so one `apply` log might involve multiple RocksDB writes.
```

>     - In other cases, bugs need to be reported.

Reviewer: ```suggestion
    - For other causes, report them as bugs.
```

> - Raft commit log is slow. TiKV Grafana's Raft I/O and commit log duration are relatively high (this metric is only available in Grafana 4.x). Each Region corresponds to an independent Raft group. Raft has a flow control mechanism, similar to the sliding window mechanism of TCP, through the parameter [raftstore] raft-max-inflight-msgs = 256 to control the size of the sliding window, if there is a hot spot Write and commit log duration is relatively high, you can moderately change the parameters, such as 1024.

Reviewer: ```suggestion
- `raft commit log` is slow. In TiKV Grafana, `Raft I/O` and `commit log duration` (only available in Grafana 4.x) metrics are relatively high. Each Region corresponds to an independent Raft group. Raft has a flow control mechanism similar to the sliding window mechanism of TCP. To control the size of a sliding window, adjust the `[raftstore] raft-max-inflight-msgs` parameter. if there is a write hotspot and `commit log duration` is high, you can properly set this parameter to a larger value, such as `1024`.
```

### I/O problem found from alarm

The cluster deployment tool (TiDB-Ansible, TiUP) is an alarm component that is deployed by default. Officials have preset related alarm items and thresholds. I/O related items include:

- TiKV_write_stall
- TiKV_raft_log_lag
- TiKV_async_request_snapshot_duration_seconds
- TiKV_async_request_write_duration_seconds
- TiKV_raft_append_log_duration_secs
- TiKV_raft_apply_log_duration_secs

## I/O problem handling plan

1. When it is confirmed as a I/O hotspot issue, you need to refer to [TiDB Hot Issue Processing] (/troubleshoot-hot-spot-issues.md) to eliminate the related I/O hotspot situation.
2. When it is confirmed that the overall I/O has reached the bottleneck, and the ability to judge the I/O from the business side will continue to keep up, then you can take advantage of the distributed database's scale capability and adopt the scheme of expanding the number of TiKV nodes to obtain greater overall I/O throughput.

> 3. Adjust some of the parameters in the above description, and use computing/memory resources in exchange for disk storage resources.

Reviewer: ```suggestion
+ Adjust some of the parameters as described above, and use computing/memory resources to make up for disk storage resources.
```

---
title: Troubleshoot High Disk I/O Usage in TiDB
summary: Learn how to locate and address the issue of high TiDB storage I/O usage.
category: reference

> aliases: ['/docs/dev/troubleshoot-high-disk-io/']

Reviewer: ```suggestion
```

---

# Troubleshoot High Disk I/O Usage in TiDB

This document introduces how to locate and address the issue of high disk I/O usage in TiDB.

## Check the current I/O metrics

If TiDB's response slows down after you have troubleshot the CPU bottleneck and the bottleneck caused by transaction conflicts, you need to check I/O metrics to help determine the current system bottleneck.

### Locate I/O issues from monitor

> The quickest way to locate I/O issues is to view the overall I/O  status from the monitor, such as the Grafana dashboard which is deployed by default by TiDB Ansible and TiUP. The dashboard panels related to I/O include **Overview**, **Node_exporter**, **Disk-Performance**.

Reviewer: ```suggestion
The quickest way to locate I/O issues is to view the overall I/O status from the monitor, such as the Grafana dashboard which is deployed by default by TiDB Ansible and TiUP. The dashboard panels related to I/O include **Overview**, **Node_exporter**, **Disk-Performance**.
```

#### The third type of monitoring panels

In **TiKV-Details** > **Storage**, there are monitoring metrics related to storage:

- `Storage command total`: Indicates the number of different commands received.
- `Storage async write duration`: Includes monitoring metrics such as `disk sync duration`, which might be related to Raft I/O. If you encounter an abnormal situation, check the working statuses of related components by checking logs.

#### Other panels

In addition, some other panel metrics might help you determine whether the bottleneck is I/O, and you can try to set some parameters. By checking the prewrite/commit/raw-put (for raw key-value clusters only) of TiKV gRPC duration, you can determine that the bottleneck is indeed the slow TiKV write. The common situations of slow TiKV writes are as follows:

- `append log` is slow. TiKV Grafana's `Raft I/O` and `append log duration` metrics are relatively high, which is often due to slow disk writes. You can check the value of `WAL Sync Duration max` in **RocksDB-raft** to determine the cause of slow `append log`. Otherwise, you might need to report a bug.

> - The `raftstore` thread is busy. In TiKV Grafana, `Raft Propose`/`propose wait duration` is significantly higher than `append log duration`. Please check the following aspects for troubleshooting:

Reviewer: ```suggestion
- The `raftstore` thread is busy. In TiKV Grafana, `Raft Propose`/`propose wait duration` is significantly higher than `append log duration`. Check the following aspects for troubleshooting:
```

    - Whether the CPU resource of the machine is insufficient.

- `append log` is slow. TiKV Grafana's `Raft I/O` and `append log duration` metrics are relatively high, which might usually occur along with relatively high `Raft Propose`/`apply wait duration`. The possible causes are as follows:
  
    - The value of `apply-pool-size` of `[raftstore]` is too small. It is recommended to set this value between `[1, 5]` and not too large. The value of `Thread CPU`/`apply cpu` is also relatively high.
    - Insufficient CPU resources on the machine.
    - Write hotspot issue of a single Region (Currently, the solution to this issue is still on the way). The CPU usage of a single `apply` thread is high (which can be viewed by modifying the Grafana expression, appended with `by (instance, name)`).
    - Slow write into RocksDB, and `RocksDB kv`/`max write duration` is high. A single Raft log might contain multiple key-value pairs (kv). 128 kvs are written to RocksDB in a batch, so one `apply` log might involve multiple RocksDB writes.
    - For other causes, report them as bugs.

> - `raft commit log` is slow. In TiKV Grafana, `Raft I/O` and `commit log duration` (only available in Grafana 4.x) metrics are relatively high. Each Region corresponds to an independent Raft group. Raft has a flow control mechanism similar to the sliding window mechanism of TCP. To control the size of a sliding window, adjust the `[raftstore] raft-max-inflight-msgs` parameter. if there is a write hotspot and `commit log duration` is high, you can properly set this parameter to a larger value, such as `1024`.

Reviewer: ```suggestion
- `raft commit log` is slow. In TiKV Grafana, `Raft I/O` and `commit log duration` (only available in Grafana 4.x) metrics are relatively high. Each Region corresponds to an independent Raft group. Raft has a flow control mechanism similar to the sliding window mechanism of TCP. To control the size of a sliding window, adjust the `[raftstore] raft-max-inflight-msgs` parameter. If there is a write hotspot and `commit log duration` is high, you can properly set this parameter to a larger value, such as `1024`.
```

    
    For example, when the pressure of `default cf compaction` is relatively high, you can change the parameter`[rocksdb.defaultcf] compression-per-level = ["no", "no", "lz4", "lz4", "lz4", "zstd" , "zstd"]`  to `compression-per-level = ["no", "no", "zstd", "zstd", "zstd", "zstd", "zstd"]`.

### I/O problem found from alarm

The cluster deployment tools (TiDB Ansible and TiUP) deploy the cluster with alert components by default that have built-in alert items and thresholds. The following alert items are related to I/O:

- TiKV_write_stall
- TiKV_raft_log_lag
- TiKV_async_request_snapshot_duration_seconds
- TiKV_async_request_write_duration_seconds
- TiKV_raft_append_log_duration_secs
- TiKV_raft_apply_log_duration_secs

## Handle I/O issues

+ When an I/O hotspot issue is confirmed to occur, you need to refer to Handle TiDB Hotspot Issues to eliminate the I/O hotspots.

> + When it is confirmed that the overall I/O performance has become the bottleneck, and you can determine that the I/O performance will keep falling behind from the application side, then you can take advantage of the distributed database's capability to scale and scale out the number of TiKV nodes to have greater overall I/O throughput.

Reviewer: ```suggestion
+ When it is confirmed that the overall I/O performance has become the bottleneck, and you can determine that the I/O performance will keep falling behind in the application side, then you can take advantage of the distributed database's capability of scaling and scale out the number of TiKV nodes to have greater overall I/O throughput.
```
